x-pack/platform/plugins/shared/alerting/server/plugin.ts

```diff
@@ -451,15 +457,38 @@ export class AlertingPlugin {
         if (!(ruleType.minimumLicenseRequired in LICENSE_TYPE)) {
           throw new Error(`"${ruleType.minimumLicenseRequired}" is not a valid license type`);
         }
+
+        // validate cancelAlertsOnTimeout if set explicitly on the rule type definition
+        if (
+          ruleType.cancelAlertsOnRuleTimeout === false &&
+          (ruleType.autoRecoverAlerts == null || ruleType.autoRecoverAlerts === true)
+        ) {
+          throw new Error(
+            `Rule type "${ruleType.id}" cannot have both cancelAlertsOnRuleTimeout set to false and autoRecoverAlerts set to true.`
+          );
+        }
+
         ruleType.ruleTaskTimeout = getRuleTaskTimeout({
           config: this.config.rules,
           ruleTaskTimeout: ruleType.ruleTaskTimeout,
           ruleTypeId: ruleType.id,
         });
-        ruleType.cancelAlertsOnRuleTimeout =
-          ruleType.cancelAlertsOnRuleTimeout ?? this.config.cancelAlertsOnRuleTimeout;
         ruleType.doesSetRecoveryContext = ruleType.doesSetRecoveryContext ?? false;
         ruleType.autoRecoverAlerts = ruleType.autoRecoverAlerts ?? true;
+
+        if (
+          ruleType.autoRecoverAlerts === true &&
+          this.config.cancelAlertsOnRuleTimeout === false
+        ) {
+          this.logger.debug(
+            `Setting xpack.alerting.cancelAlertsOnRuleTimeout=false is incompatible with rule type "${ruleType.id}" and will be ignored.`
+          );
+          ruleType.cancelAlertsOnRuleTimeout = true;
+        } else {
+          ruleType.cancelAlertsOnRuleTimeout =
+            ruleType.cancelAlertsOnRuleTimeout ?? this.config.cancelAlertsOnRuleTimeout;
+        }
+
```

Comment:
Feels like we need to add `ruleTypeRegistry.validate` and move all these checks to there but not concern of this PR.